Allow non-alphabetic characters in expression generator

mfojtik opened the pull request:
This allows to generate stronger passwords

mfojtik commented:
@bparees @smarterclayton

bparees commented on pkg/template/generator/expressionvalue.go:
```diff
@@ -28,15 +28,16 @@ type ExpressionValueGenerator struct {
 	rangeExp      = regexp.MustCompile(`([\\]?[a-zA-Z0-9]\-?[a-zA-Z0-9]?)`)
 	generatorsExp = regexp.MustCompile(`\[([a-zA-Z0-9\-\\]+)\](\{([0-9]+)\})`)
-	expressionExp = regexp.MustCompile(`\[(\\w|\\d|\\a)|([a-zA-Z0-9]\-[a-zA-Z0-9])+\]`)
+	expressionExp = regexp.MustCompile(`\[(\\w|\\d|\\a|\\S)|([a-zA-Z0-9]\-[a-zA-Z0-9])+\]`)
```
what is this doing: ([a-zA-Z0-9]\-[a-zA-Z0-9])+
?
shouldn't it just match [\w\d\S]+ ?

mfojtik replied:
@bparees I think we are listing what 'wildchars' are allowed explicitly so you can have [abc\S]{3} which should tell generator to generate random string that includes "a,b,c" + all non-alphabetic characters

bparees commented:
lgtm.  will need to wait for post 1.1.

mfojtik commented:
@rnetuka FYI

liggitt commented on pkg/template/generator/expressionvalue_test.go:
```diff
@@ -16,8 +16,10 @@ func TestExpressionValueGenerator(t *testing.T) {
 		{"[\\d]{4}", "6841"},
 		{"[\\w]{20}", "UDd5PEJLyHZ7VpdMGC78"},
 		{"[\\a]{10}", "fUx4sLZWES"},
-		{"admin[0-9]{2}[A-Z]{2}", "admin05XX"},
```
move the `generator` construction and seeding inside the loop so that changes don't affect all following tests

liggitt commented on pkg/template/generator/expressionvalue.go:
```diff
@@ -94,6 +95,8 @@ func replaceWithGenerated(s *string, expression string, ranges [][]byte, length
 			alphabet += Numerals
 		case `\a`:
 			alphabet += Alphabet + Numerals
+		case `\S`:
```
`\S` normally means non-whitespace in regex... why are we using that to mean "Non-Alphabet" here? that's unexpected

mfojtik replied:
@liggitt i'm open for suggestions... I thought \A might be better, in regexp that will mean negation of \a which we use for aphabet + numbers

smarterclayton commented:
API change approved

liggitt commented on pkg/template/generator/expressionvalue.go:
```diff
@@ -94,6 +95,8 @@ func replaceWithGenerated(s *string, expression string, ranges [][]byte, length
 			alphabet += Numerals
 		case `\a`:
 			alphabet += Alphabet + Numerals
```
appending the same alphabet multiple times means we can unexpected weight generated values... for example `[\w\d\a]` will include numbers three times, and alphabet twice... do we care about that?

mfojtik replied:
yeah, we should remove duplicated characters from the alphabet

bparees replied:
\a probably should have been alphabetic characters only (not numeric)

bparees replied:
barring the ability to change that at this point, @liggitt is probably right, we should just add each "group" to the pool of potential characters once.

mfojtik replied:
that will be breaking api change, but I agree, the \a should be alphabet only

bparees replied:
yes probably too late now.

bparees commented on pkg/template/generator/expressionvalue.go:
```diff
@@ -28,15 +28,16 @@ type ExpressionValueGenerator struct {
-	ASCII    = Alphabet + Numerals + "~!@#$%^&*()-_+={}[]\\|<,>.?/\"';:`"
+	Alphabet    = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+	Numerals    = "0123456789"
+	NonAlphabet = "~!@#$%^&*()-_+={}[]\\|<,>.?/\"';:`"
```
Symbols.

Alphabet is A-Z only, it doesn't include or imply numbers.

mfojtik commented:
@liggitt @bparees fixed:

* add function that removes all duplicated characters from the alphabet
* add test for that function
* renamed `\S` to `\A`
* fixed generator test

bparees commented:
lgtm

mfojtik commented:
[test]

mfojtik commented:
@bparees merge?

bparees commented:
[merge]!
